pkg/distributor/distributor_test.go

```diff
@@ -2593,6 +2593,98 @@ func TestDistributor_MetricsForLabelMatchers(t *testing.T) {
 	}
 }
 
+func TestDistributor_MetricsForLabelMatchers_adjustPushDownLimit(t *testing.T) {
+	const numIngesters = 10
+
+	now := model.Now()
+
+	fixtures := []struct {
+		lbls      []mimirpb.LabelAdapter
+		value     float64
+		timestamp int64
+	}{
+		{labelAdapters(labels.MetricName, "test_1", "status", "2xx"), 1, 100000},
+		{labelAdapters(labels.MetricName, "test_1", "status", "3xx"), 1, 110000},
+		{labelAdapters(labels.MetricName, "test_1", "status", "5xx"), 1, 120000},
+	}
+
+	tests := map[string]struct {
+		setupFunc     func(*prepConfig)
+		hints         *storage.SelectHints
+		expectedLimit int
+	}{
+		"should adjust push-down limit": {
+			hints: &storage.SelectHints{
+				Limit: 50,
+			},
+			expectedLimit: 17,
+		},
+		"should adjust push-down limit if shuffle shard size is set": {
+			setupFunc: func(testConfig *prepConfig) {
+				testConfig.shuffleShardSize = 6
+			},
+			hints: &storage.SelectHints{
+				Limit: 50,
+			},
+			expectedLimit: 25, // the limit is divided between two shards (ingest_shard_size=6; RF=3)
+		},
+		"should adjust push-down limit if ingest-storage is enabled": {
+			setupFunc: func(testConfig *prepConfig) {
+				testConfig.ingestStorageEnabled = true
+				testConfig.limits = prepareDefaultLimits()
+				testConfig.limits.IngestionPartitionsTenantShardSize = 3
```

Comment:
I suggest a test with ingest storage enabled and IngestionPartitionsTenantShardSize=0 (disabled).